services/src/pro/datasets/in_memory.rs

```diff
@@ -266,15 +350,21 @@ impl DatasetProvider<UserSession> for ProHashMapDatasetDb {
 
     async fn load(&self, session: &UserSession, dataset: &DatasetId) -> Result<Dataset> {
         ensure!(
-            self.dataset_permissions
+            self.backend
+                .read()
+                .await
```

Comment:
I think you should only get the lock once.

Reply:
fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1